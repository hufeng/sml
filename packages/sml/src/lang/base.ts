import path from 'node:path'
import { exec } from 'node:child_process'
import Builder from '../common/builder'
import {
  ActorStyleType,
  BaseAst,
  DirectionType,
  GlobalConfigType,
  PackageStyle,
} from './types'

export const noop = () => {}

// ~~~~~~~~~~ builder ~~~~~~~~~~~~~~~~~~~~~~
class ConfigBuilder {
  private config: GlobalConfigType

  constructor(config: GlobalConfigType) {
    this.config = config
  }

  /**
   * setting actor style
   * @param style
   * @returns
   */
  actorStyle(style: ActorStyleType = 'default') {
    this.config.actorStyle = style
    return this
  }

  /**
   * setting package style
   * @param style
   * @returns
   */
  packageStyle(style: PackageStyle = 'Rectangle') {
    this.config.packageStyle = style
    return this
  }

  /**
   * setting direction
   * @param direction
   */
  direction(direction: DirectionType) {
    this.config.direction = direction
  }

  theme(theme: string = 'cerulean-outline') {
    this.config.theme = theme
  }
}

// ~~~~~~~~~~ lang base ~~~~~~~~~~~~~~~~

/**
 * all lang base class
 */
export class Lang<T extends { title: string; config?: GlobalConfigType }> {
  protected meta: T

  constructor(meta: T) {
    this.meta = meta
    this.meta.config = {
      actorStyle: 'default',
      direction: 'left->right',
      packageStyle: 'Rectangle',
      theme: 'cerulean-outline',
    }
  }

  get configuration() {
    return new ConfigBuilder(this.meta.config!)
  }
}

/**
 * all Emitter base class
 */
export abstract class Emitter<T extends BaseAst> {
  protected meta: T
  protected s: Builder

  constructor(meta: T) {
    this.meta = meta
    this.s = new Builder()
  }

  abstract emitCode(): string

  protected emitConfig() {
    const { config } = this.meta
    // setting actor style
    this.s
      .ifStr(
        config!.actorStyle !== 'default',
        `skinparam actorStyle ${config!.actorStyle}`,
      )
      // setting direction
      .str(`${config!.direction.replace('->', ' to ')} direction\n`)
    // FIXME lose package style
  }

  plantUML(img: string) {
    const jar = path.join(__dirname, '../../bin/plantuml-1.2022.7.jar')

    exec(
      `${this.emitCode()} | java -jar ${jar} -pipe > ${img}`,
      (err, stdout, stderr) => {
        if (err) {
          throw err
        }
        console.log(stdout)
        console.log(stderr)
      },
    )
  }
}
