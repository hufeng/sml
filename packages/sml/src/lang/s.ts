export default class S {
  private buff: Array<string | S>

  constructor() {
    this.buff = []
  }

  $s(s: string) {
    this.buff.push(s)
    return this
  }

  $fn(fn: (s: S) => void) {
    fn(this)
    return this
  }

  $if(cond: boolean, str: string) {
    if (cond) {
      this.buff.push(str)
    }
    return this
  }

  $for<T>(arr: Array<T> = [], fn: (s: S, v: T, i?: number) => void) {
    if (arr.length === 0) {
      return this
    }

    arr.forEach((v, i) => {
      fn(this, v, i)
    })
    return this
  }

  $fors<T>(arr: Array<T> = [], fn: (s: S, v: T, i?: number) => void) {
    if (arr.length === 0) {
      return this
    }

    this.buff.push('')
    // append for
    arr.forEach((v, i) => {
      fn(this, v, i)
    })
    this.buff.push('')

    return this
  }

  reset() {
    this.buff = []
    return this
  }

  toString(delimiter = '\n') {
    return this.buff.join(delimiter)
  }
}
