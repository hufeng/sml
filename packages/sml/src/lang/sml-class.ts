import { Lang } from './sml'
import { globalCollections } from './sml-global'

// ~~~~~~~~~~~~~~ basic type ~~~~~~~~~~~~~~~~~
type DataType = string
export type VisibleType = 'private' | 'protected' | 'public' | 'package_private'
type ParamType = { name: string; type: DataType }
type MethodOptional = (m: MethodType) => void
type VisibleOptional = (v: { visible: VisibleType }) => void
type abstractOptional = (a: { abstract: boolean }) => void

// ~~~~~~~~~~~~~~ container type ~~~~~~~~~~~~
type FiledType = {
  name: string
  type: DataType
  visible: VisibleType
}
type MethodType = {
  abstract: boolean
  visible: VisibleType
  name: string
  params: Array<ParamType>
  ret: DataType
}
type ClazzType = {
  name: string
  abstract: boolean
  fields: Array<FiledType>
  methods: Array<MethodType>
  extends: Array<string>
  implements: Array<string>
}
type InterfaceType = {
  name: string
  implements: Array<string>
  methods: Array<MethodType>
}
type EnumFieldTyep = {
  name: string
  value?: number | string
}
type EnumType = {
  name: string
  fields: Array<EnumFieldTyep>
}
type StructType = ClazzType
type ProtocolType = InterfaceType
export interface SmlClazzMeta {
  clazzes: Array<ClazzType>
  interfaces: Array<InterfaceType>
  enums: Array<EnumType>
  structs: Array<StructType>
  protocols: Array<ProtocolType>
}

//~~~~~~~~~~~~ builder ~~~~~~~~~~~~~~~~
class ClazzBuilder {
  constructor(private clazz: ClazzType) {}

  field(name: string, type: DataType, visible: VisibleType = 'private') {
    this.clazz.fields.push({ name, type, visible })
    return this
  }

  method(
    name: string,
    params: MethodOptional,
    ret: MethodOptional,
    ...rest: Array<VisibleOptional | abstractOptional>
  ) {
    const method = {
      abstract: false,
      visible: 'public' as VisibleType,
      name,
      params: [] as Array<ParamType>,
      ret: '' as DataType,
    }
    params(method)
    ret(method)
    rest.forEach((opt) => opt(method))
    this.clazz.methods.push(method)
    return this
  }

  extends(...names: Array<string>) {
    this.clazz.extends = [...this.clazz.extends, ...names]
    return this
  }

  implements(...names: Array<string>) {
    this.clazz.implements = [...this.clazz.implements, ...names]
    return this
  }
}

class InfBuilder {
  constructor(private interfaces: InterfaceType) {}

  method(
    name: string,
    args: MethodOptional = (m: MethodType) => (m.params = []),
    ret: MethodOptional = (m: MethodType) => (m.ret = 'void'),
  ) {
    const method = {
      name,
      visible: 'public' as VisibleType,
      abstract: false,
      params: [] as Array<ParamType>,
      ret: '' as DataType,
    }
    args(method)
    ret(method)
    this.interfaces.methods.push(method)
    return this
  }

  implements(...names: Array<string>) {
    this.interfaces.implements = [...this.interfaces.implements, ...names]
    return this
  }
}

class EnumBuilder {
  constructor(private e: EnumType) {}

  field(name: string, value?: string | number) {
    this.e.fields.push({ name, value })
    return this
  }
}

// ~~~~~~~~~~~~~ define class lang modeing ~~~~~~~~~~~~~~~~~~~
enum T {
  void = 'void',
  string = `string`,
  boolean = `boolean`,
  number = `number`,

  go_int8 = `go_int8`,
  go_int16 = `go_int16`,
  go_int32 = `go_int32`,
  go_int = `go_int`,
  go_uint = `go_uint`,
  go_byte = 'go_byte',
  go_int64 = `go_int64`,
  go_uint8 = `go_uint8`,
  go_uint16 = `go_uint16`,
  go_uint32 = `go_uint32`,
  go_uint64 = `go_uint64`,
  go_float32 = `go_float32`,
  go_float64 = `go_float64`,

  py_int = 'py_int',
  py_bool = `py_bool`,
  py_float = `py_float`,

  java_byte = `java_byte`,
  java_short = `java_short`,
  java_int = `java_int`,
  java_long = `java_long`,
  java_float = `java_float`,
  java_double = `java_double`,
  java_boolean = `java_boolean`,
  java_char = `java_char`,
}
export class SmlClazzLang extends Lang {
  private meta: SmlClazzMeta

  constructor(title: string) {
    super(title)
    this.meta = {
      clazzes: [],
      interfaces: [],
      enums: [],
      structs: [],
      protocols: [],
    }
  }

  /**
   * define abstract class
   * @param name abstract class name
   * @returns
   */
  abstractClazz(name: string) {
    const clazz = {
      name,
      abstract: true,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    }
    this.meta.clazzes.push(clazz)
    return new ClazzBuilder(clazz)
  }

  /**
   * define class
   * @param name class name
   * @returns
   */
  clazz(name: string) {
    const clazz = {
      name,
      abstract: false,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    }
    this.meta.clazzes.push(clazz)
    return new ClazzBuilder(clazz)
  }

  /**
   * define interface
   * @param name interface name
   * @returns
   */
  interface(name: string) {
    const inf = {
      name,
      implements: [],
      methods: [],
    } as InterfaceType
    this.meta.interfaces.push(inf)
    return new InfBuilder(inf)
  }

  /**
   * define enum
   * @param name enum name
   * @returns
   */
  enum(name: string) {
    const e = {
      name,
      fields: [],
    }
    this.meta.enums.push(e)
    return new EnumBuilder(e)
  }

  /**
   * define struct
   * @param name struct name
   * @returns
   */
  struct(name: string) {
    const s = {
      name,
      abstract: false,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    } as ClazzType
    this.meta.structs.push(s)
    return new ClazzBuilder(s)
  }

  /**
   * define a protocol
   * @param name protocol name
   * @returns
   */
  protocol(name: string) {
    const prot = {
      name,
      implements: [],
      methods: [],
    } as InterfaceType
    this.meta.protocols.push(prot)
    return new InfBuilder(prot)
  }

  /**
   * args optional
   * @param args
   * @returns
   */
  args(...args: Array<ParamType>) {
    return (m: MethodType) => (m.params = [...m.params, ...args])
  }

  /**
   * arg optional
   */
  arg(name: string, type: string) {
    return { name, type } as ParamType
  }

  /**
   * return value type
   * @param type
   * @returns
   */
  ret(type: string) {
    return (m: MethodType) => (m.ret = type)
  }

  /**
   * setting field or method visible
   * @param v
   * @returns
   */
  visible(v: VisibleType) {
    return (val: { visible: VisibleType }) => (val.visible = v)
  }

  /**
   * setting method whether is abstract
   * @param v
   * @returns
   */
  abstract(v: boolean) {
    return (val: { abstract: boolean }) => (val.abstract = v)
  }

  t = T
}

/**
 * ClassDiagram factory
 * @param title
 * @param fn
 * @returns
 */
export function ClassDiagram(title: string, fn: (ml: SmlClazzLang) => void) {
  const lang = new SmlClazzLang(title)
  fn(lang)
  globalCollections.add(lang)
  return lang
}
