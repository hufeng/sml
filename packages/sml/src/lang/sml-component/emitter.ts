import { Emitter } from '../base'
import { SmlComponentAst } from '../types'

export class PumlComponentEmitter extends Emitter<SmlComponentAst> {
  emitCode() {
    const {
      packages,
      nodes,
      clouds,
      databases,
      components,
      infs,
      links,
      vlinks,
      rels,
    } = this.meta

    return this.s
      .str('@startuml')
      .forStr(
        packages,
        (s, v) =>
          s
            .str(`  package "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        packages.length > 0 ? '\n' : '',
      )
      .forStr(
        nodes,
        (s, v) =>
          s
            .str(`  node "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `   component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str(' }'),
        nodes.length > 0 ? '\n' : '',
      )
      .forStr(
        clouds,
        (s, v) =>
          s
            .str(`  cloud "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        clouds.length > 0 ? '\n' : '',
      )
      .forStr(
        databases,
        (s, v) =>
          s
            .str(`  database "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        databases.length > 0 ? '\n' : '',
      )
      .forStr(components, (s, v) =>
        s.str(`  component "${v.title}"${v.name ? '  as ' + v.name : ''}`),
      )
      .forStr(
        infs,
        (s, v) =>
          s.str(`  interface "${v.title}"${v.name ? '  as ' + v.name : ''}`),
        infs.length > 0 ? '\n' : '',
      )
      .forStr(links, (s, v) => s.str(`  ${v.from} --> ${v.to}`))
      .forStr(vlinks, (s, v) => s.str(`  ${v.from} ..> ${v.to}`))
      .forStr(rels, (s, v) => s.str(`  ${v.from} - ${v.to}`))
      .str('@enduml')
      .toString('\n')
  }
}
