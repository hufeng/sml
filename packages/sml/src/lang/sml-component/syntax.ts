import { Lang } from '../base'
import { ComponentContainer, SmlComponentAst } from '../types'

class ContainerBuilder {
  constructor(private meta: ComponentContainer) {}

  interface(label: string, name?: string) {
    this.meta.infs.push({ label, name })
    return this
  }

  component(label: string, name?: string) {
    this.meta.components.push({ label, name })
    return this
  }
}

export class SmlComponentLang extends Lang<SmlComponentAst> {
  constructor(meta: SmlComponentAst) {
    super(meta)
  }

  #container(
    label: string,
    collections: 'packages' | 'nodes' | 'databases' | 'clouds',
    fn?: (c: ContainerBuilder) => void,
  ) {
    const ast = {
      label,
      components: [],
      infs: [],
    }
    fn && fn(new ContainerBuilder(ast))
    this.meta[collections].push(ast)
    return this
  }

  package(label: string, fn?: (p: ContainerBuilder) => void) {
    return this.#container(label, 'packages', fn)
  }

  node(label: string, fn?: (n: ContainerBuilder) => void) {
    return this.#container(label, 'nodes', fn)
  }

  database(label: string, fn?: (d: ContainerBuilder) => void) {
    return this.#container(label, 'databases', fn)
  }

  /**
   * 生成一个☁️包裹的作用域
   * @param label
   * @param fn
   * @returns
   */
  cloud(label: string, fn?: (c: ContainerBuilder) => void) {
    return this.#container(label, 'clouds', fn)
  }

  component(label: string, name?: string) {
    this.meta.components.push({ label, name })
    return this
  }

  interface(label: string, name?: string) {
    this.meta.infs.push({ label, name })
    return this
  }

  // ~~~~~~~~~ relations ~~~~~~~~~~~~

  link(from: string, to: string) {
    this.meta.links.push({ from, to })
    return this
  }

  vlink(from: string, to: string) {
    this.meta.vlinks.push({ from, to })
    return this
  }

  rel(from: string, to: string) {
    this.meta.rels.push({ from, to })
    return this
  }
}
