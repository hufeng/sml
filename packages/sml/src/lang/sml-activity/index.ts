import { PumlActivityEmitter } from './emitter'
import { SmlActivityLang } from './syntax'

export function SmlActivityDiagram(
  title: string,
  fn: (ml: SmlActivityLang) => void,
) {
  const ast = {
    title,
    components: [],
    links: [],
    vlinks: [],
    notes: [],
  }

  fn(new SmlActivityLang(ast))

  return { ast, emitter: new PumlActivityEmitter(ast) }
}
