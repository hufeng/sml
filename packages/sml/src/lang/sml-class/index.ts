import { PumlClassEmitter } from './emitter'
import { SmlClazzLang as SmlClassLang } from './syntax'

/**
 * ClassDiagram factory
 * @param title
 * @param fn
 * @returns
 */
export function ClassDiagram(title: string, fn: (ml: SmlClassLang) => void) {
  const ast = {
    title,
    clazzes: [],
    interfaces: [],
    enums: [],
    structs: [],
    protocols: [],
  }

  fn(new SmlClassLang(ast))

  return {
    ast,
    emitter: new PumlClassEmitter(ast),
  }
}
