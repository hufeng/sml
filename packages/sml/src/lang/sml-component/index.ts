import { PumlComponentEmitter } from './emitter'
import { SmlComponentLang } from './syntax'

export function ComponentDiagram(
  title: string,
  fn: (ml: SmlComponentLang) => void,
) {
  const ast = {
    title,

    zones: [],
    nodes: [],
    databases: [],
    clouds: [],

    components: [],
    infs: [],

    notes: [],
    vlinks: [],
    links: [],
    rels: [],
  }
  fn(new SmlComponentLang(ast))
  return { ast, emitter: new PumlComponentEmitter(ast) }
}
