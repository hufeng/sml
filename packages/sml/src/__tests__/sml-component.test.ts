import { ComponentDiagram } from '../lang/sml-component'

describe('component diagram test suites', () => {
  it('test all components', () => {
    const { ast, emitter } = ComponentDiagram('component title', (ml) => {
      // elements
      const cloud = ml.zone('cloud').type('cloud')
      const c1 = ml.component('Nodejs').belongTo(cloud)
      const i1 = ml.interface('http').belongTo(cloud)

      const zone = ml.zone('bff')
      const c2 = ml.component('service-a').belongTo(zone)
      const c3 = ml.component('service-b').belongTo(zone)
      const i2 = ml.interface('rpc').belongTo(zone)

      const db = ml.zone('DataBase').type('database')
      const c4 = ml.component('mysql')
      const c5 = ml.component('redis')
      db.has(c4, c5)

      const node = ml.zone('Node').type('node')
      const c6 = ml.component('Go')
      node.has(c6)

      // links
      c1.link(c2, (l) => l.noteOf('connected'))
      c2.link(c3)
      c3.link(c4)
      c4.vlink(c5)
      c5.vlink(c6)

      i1.rel(c1)
      i2.rel(c3)
    })
    expect(ast).toMatchInlineSnapshot(`
      {
        "components": [],
        "config": {
          "actorStyle": "default",
          "direction": "left->right",
          "packageStyle": "Rectangle",
          "theme": "sketchy-outline",
        },
        "infs": [],
        "links": [
          {
            "from": "c_4b00bbf4",
            "note": {
              "label": "connected",
              "position": "right",
            },
            "to": [
              "c_7fdd65ec",
            ],
          },
          {
            "from": "c_7fdd65ec",
            "to": [
              "c_d77ba138",
            ],
          },
          {
            "from": "c_d77ba138",
            "to": [
              "c_81c3b080",
            ],
          },
        ],
        "notes": [],
        "rels": [
          {
            "from": "i_80791b3a",
            "to": [
              "c_4b00bbf4",
            ],
          },
          {
            "from": "i_da0fb2ac",
            "to": [
              "c_d77ba138",
            ],
          },
        ],
        "title": "component title",
        "vlinks": [
          {
            "from": "c_81c3b080",
            "to": [
              "c_86a1b907",
            ],
          },
          {
            "from": "c_86a1b907",
            "to": [
              "c_5f075ae3",
            ],
          },
        ],
        "zones": [
          {
            "components": [
              {
                "id": "c_4b00bbf4",
                "label": "Nodejs",
              },
            ],
            "infs": [
              {
                "id": "i_80791b3a",
                "label": "http",
              },
            ],
            "label": "cloud",
            "name": "z_a1234b31",
            "type": "cloud",
          },
          {
            "components": [
              {
                "id": "c_7fdd65ec",
                "label": "service-a",
              },
              {
                "id": "c_d77ba138",
                "label": "service-b",
              },
            ],
            "infs": [
              {
                "id": "i_da0fb2ac",
                "label": "rpc",
              },
            ],
            "label": "bff",
            "name": "z_18452e2c",
            "type": "package",
          },
          {
            "components": [
              {
                "id": "c_81c3b080",
                "label": "mysql",
              },
              {
                "id": "c_86a1b907",
                "label": "redis",
              },
            ],
            "infs": [],
            "label": "DataBase",
            "name": "z_c55bea99",
            "type": "database",
          },
          {
            "components": [
              {
                "id": "c_5f075ae3",
                "label": "Go",
              },
            ],
            "infs": [],
            "label": "Node",
            "name": "z_6c3a6944",
            "type": "node",
          },
        ],
      }
    `)
    expect(emitter.emitCode()).toMatchSnapshot()
  })
})
