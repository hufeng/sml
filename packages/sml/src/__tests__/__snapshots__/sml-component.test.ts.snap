// Vitest Snapshot v1

exports[`component diagram test suites > test all components 1`] = `
"@startuml
!theme cerulean-outline

package \\"bff\\" {
  component \\"service-a\\"  as c2
  component \\"service-b\\"  as c3
  interface \\"http\\"  as i2
}

node \\"Node\\" {
  component \\"Go\\"  as c6
}

cloud \\"cloud\\" {
  component \\"Nodejs\\"  as c1
  interface \\"http\\"  as i1
}

database \\"Database\\" {
  component \\"mysql\\"  as c4
  component \\"redis\\"  as c5
}

c1 --> c2
c2 --> c3
c3 ..> c4
c4 ..> c5
c5 ..> c6
i1 - c1
i2 - c3
@enduml"
`;
